criu/sk-tcp.c

```diff
@@ -428,7 +430,7 @@ int restore_one_tcp(int fd, struct inet_sk_info *ii)
 
 	pr_info("Restoring TCP connection\n");
 
-	if (opts.tcp_close && ii->ie->state != TCP_LISTEN && ii->ie->state != TCP_CLOSE) {
+	if (dump_with_tcp_close | (opts.tcp_close && ii->ie->state != TCP_LISTEN && ii->ie->state != TCP_CLOSE)) {
 		if (shutdown(fd, SHUT_RDWR) && errno != ENOTCONN) {
```

Comment:
The problem occurs because when the `--tcp-close` option is used during checkpoint CRIU does not dump sockets in `TCP_LISTEN` and `TCP_CLOSE`.

Looking closer at the commit message in https://github.com/checkpoint-restore/criu/commit/2c37042821906f013634e305899cb25ff1a5a7b1:
```
New restore option 'tcp-close' was introduced. It restores all connected
TCP sockets in TCP_CLOSE state. Here we consider tcp sockets in
TCP_ESTABLISHED, TCP_FIN_WAIT2, TCP_FIN_WAIT1, TCP_CLOSE_WAIT,
TCP_LAST_ACK, TCP_CLOSING, TCP_SYN_SENT states as connected sockets.
This is consistent with current CRIU usage of these states. Thus this
option doesn't affect sockets with original states of TCP_LISTEN and
TCP_CLOSE.
```

and https://github.com/checkpoint-restore/criu/commit/e42f5e032e0a518374f8e38aae6af198d2b66da8:

```
In this case, states of established tcp connections will not be dumped
and they will not be blocked. This will be useful in case of snapshots,
when we don't need to restore tcp connections.
```

It looks like the intended behaviour is for sockets in `TCP_LISTEN` and `TCP_CLOSE` to be checkpointed and restored when `--tcp-closed` is used. The following change could fix this:

```diff
diff --git a/criu/sk-tcp.c b/criu/sk-tcp.c
index 8e78fca8c..2f64f0f40 100644
--- a/criu/sk-tcp.c
+++ b/criu/sk-tcp.c
@@ -240,7 +240,7 @@ int dump_one_tcp(int fd, struct inet_sk_desc *sk, SkOptsEntry *soe)
        if (sk->dst_port == 0)
                return 0;
 
-       if (opts.tcp_close) {
+       if (opts.tcp_close && sk->state != TCP_LISTEN && sk->state != TCP_CLOSE) {
                return 0;
        }
```

Reply:
@avagin what do you think?

Reply:
I am not sure that you understand what you mean. We check that sk->dst_port isn't 0, it means that this is an established connection. Why do we need to check TCP_LISTEN and  TCP_CLOSE here?

Reply:
> Why do we need to check TCP_LISTEN and TCP_CLOSE here?

You're right, only check for `TCP_CLOSE` is necessary here. I've updated the patch in the comment above.

Reply:
The commit messages that you mentioned do explain why we may want to dump closed sockets. Do you remember why we decided to add --tcp-close? If we want to make a snapshot, we don't care about tcp connections, because when we will restore processes other sides of tcp connections will not be there. The second reason is that we don't want to leave any iptables or nftables rules that we use to lock connections.